fdbclient/ManagementAPI.actor.cpp

```diff
@@ -3786,11 +3761,56 @@ ACTOR Future<Void> takeExclusiveReadLockOnRange(Transaction* tr, KeyRange range,
 ACTOR Future<Void> releaseExclusiveReadLockOnRange(Transaction* tr, KeyRange range, RangeLockOwnerName ownerUniqueID) {
 	tr->setOption(FDBTransactionOptions::LOCK_AWARE);
 	tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-	wait(prepareExclusiveRangeLockOperation(tr, range, ownerUniqueID));
+	wait(prepareExclusiveRangeUnlockOperation(tr, range, ownerUniqueID));
 	// At this point, no lock presents on the range.
 	// Unlock by overwiting the range.
 	wait(krmSetRangeCoalescing(tr, rangeLockPrefix, range, normalKeys, rangeLockStateSetValue(RangeLockStateSet())));
-	TraceEvent(SevInfo, "ReleaseExclusiveReadLockOnRange").detail("Range", range);
+	TraceEvent(SevInfo, "ReleaseExclusiveReadLockTransactionOnRange").detail("Range", range);
+	return Void();
+}
+
+ACTOR Future<Void> releaseExclusiveReadLockByUser(Database cx, RangeLockOwnerName ownerUniqueID) {
+	state Key beginKey = normalKeys.begin;
+	state Key endKey = normalKeys.end;
+	state Transaction tr(cx);
+	state int i = 0;
+	state RangeResult result;
+	state KeyRange rangeToRead;
+	state RangeLockStateSet currentRangeLockStateSet;
+	state KeyRange currentRange;
+	loop {
+		rangeToRead = Standalone(KeyRangeRef(beginKey, endKey));
+		try {
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			result.clear();
+			wait(store(result, krmGetRanges(&tr, rangeLockPrefix, rangeToRead)));
+			i = 0;
+			for (; i < result.size() - 1; i++) {
+				currentRange = KeyRangeRef(result[i].key, result[i + 1].key);
+				if (result[i].value.empty()) {
+					beginKey = currentRange.end;
+					continue;
+				}
+				currentRangeLockStateSet = decodeRangeLockStateSet(result[i].value);
+				ASSERT(currentRangeLockStateSet.isValid());
+				if (currentRangeLockStateSet.isLockedFor(RangeLockType::ExclusiveReadLock) &&
+				    currentRangeLockStateSet.getAllLockStats()[0].getOwnerUniqueId() == ownerUniqueID) {
+					wait(krmSetRangeCoalescing(
+					    &tr, rangeLockPrefix, currentRange, normalKeys, rangeLockStateSetValue(RangeLockStateSet())));
+					wait(tr.commit());
```

Comment:
this is a bit weird: essentially you create a txn commit for each range, instead of updating many ranges in one transaction. If the number of ranges is large, this is bad for performance.

Reply:
This is because the limitation of krmSetRangeCoalescing
```
// Sets a range of keys in a key range map, coalescing with adjacent regions if the values match
// Ranges outside of maxRange will not be coalesced
// CAUTION: use care when attempting to coalesce multiple ranges in the same prefix in a single transaction
```

Reply:
Please add a TODO here.